### PR TITLE
feat(desktop/contacts) Improvements in contacts

### LINKED ISSUE #4100: [refactoring/base_02] Contact list is empty in Profile
# Bug Report

## Description

Contact list is empty in Profile

## Steps to reproduce

1. open app
2. open 1x1 chat and add someone to contacts, like any random public key `0x0478632dbe2f720f1f7f1ade080010fbede8ed1f4eebc164e2621bb55bed45757f4a3eb0495552cc26480f209607aeb34223fc4d8363fb074fede6082baf26db08`
3. open profile -> contacts

#### Expected behavior

List of contacts contain recently added contact

#### Actual behavior

List of contacts say that no contacts added

Screenshot 2021-11-11 at 20 48 46

Screenshot 2021-11-11 at 20 48 56

### Additional Information

- Status desktop version: r`efactoring_base02` branch, commit https://github.com/status-im/status-desktop/commit/224cb706623c84e3573010c41d435bdea19ffa19
- Operating System: masOS Big Sur

### PR DESCRIPTION
**Pull request by alexandraB99**
Closes #4100

### What does the PR do
Bring contacts updates currently in base_bc in master

### Affected areas
Contacts all over

### Screenshot of functionality
https://user-images.githubusercontent.com/31625338/143152584-c8085e4d-499e-484d-90f1-2ae25a1b8d80.mov

### Cool Spaceship Picture
ITS_Interplanetary_Spaceship,_landed_on_Enceladus

### CODE REVIEW

File: src/app/modules/main/profile_section/contacts/controller.nim [hunk 1]
```diff
@@ -16,9 +17,6 @@ type
     contactsService: contacts_service.Service
     accountsService: accounts_service.ServiceInterface
 
-# forward declaration:
-method getContacts*[T](self: Controller[T]): seq[ContactsDto]
```
Comment (PascalPrecht) [inline src/app/modules/main/profile_section/contacts/controller.nim:20]: Any reason this was removed? I assume this was introduced for a reason as well...

File: src/app/modules/main/profile_section/contacts/models/contact_list.nim [hunk 2]
```diff
@@ -116,23 +115,38 @@ QtObject:
...
+  # It also may has an impact to the app performances since it's handled on multiple places on the qml side.
+  # Would be good to get rid of it durign refactor phase.
   proc contactChanged*(self: ContactList, pubkey: string) {.signal.}
 
   proc updateContact*(self: ContactList, contact: ContactsDto) =
     var found = false
     let topLeft = self.createIndex(0, 0, nil)
     let bottomRight = self.createIndex(self.contacts.len, 0, nil)
-    for c in self.contacts:
+    for c in self.contacts.mitems:
```
Comment (PascalPrecht) [inline src/app/modules/main/profile_section/contacts/models/contact_list.nim:149]: Has the data structure changed so that we now need to access `mitems` instead of just `contacts`?

File: src/app_service/service/contacts/service.nim [hunk 3]
```diff
@@ -60,129 +71,99 @@ QtObject:
...
-    # let nickname =
-    #   if (nicknameToSet == ""):
-    #     contact.localNickname
-    #   elif (nicknameToSet == DELETE_CONTACT):
-    #     ""
-    #   else:
-    #     nicknameToSet
-
-    # contact.localNickname = nickname
+    self.events.emit(SIGNAL_CONTACT_REMOVED, ContactArgs(contactId: contact.id))
```
Comment (PascalPrecht) [inline src/app_service/service/contacts/service.nim:133]: This will probably need an update once `base_02` is rebased on top of `master`.

Using `saveContact` is no longer correct here as `wakuexit_saveContact` has been removed from status-go. In other words, instead of calling `saveContact` for every contact related action (blocking/unblocking/adding/removing etc), there's now dedicated APIs.

So this code will break contacts functionality all over again.

File: src/app_service/service/contacts/service.nim [hunk 4]
```diff
@@ -60,129 +71,99 @@ QtObject:
...
+
+  proc changeContactNickname*(self: Service, publicKey: string, nickname: string) =
+    var contact = self.getContactById(publicKey)
+    contact.localNickname = nickname
+
     self.saveContact(contact)
```
Comment (PascalPrecht) [inline src/app_service/service/contacts/service.nim:139]: Same here, we now have a dedicated API for changing the local nickname

File: src/app_service/service/contacts/service.nim [hunk 5]
```diff
@@ -60,129 +71,99 @@ QtObject:
...
 
   proc unblockContact*(self: Service, publicKey: string) =
-    # status_contacts.unblockContact(publicKey)
-    var contact = self.getContact(publicKey)
+    var contact = self.getContactById(publicKey)
     contact.blocked = false
+
     self.saveContact(contact)
-    self.events.emit("contactUnblocked", old_status_contacts.ContactIdArgs(id: publicKey))
+    self.events.emit(SIGNAL_CONTACT_UNBLOCKED, ContactArgs(contactId: contact.id))
```
Comment (PascalPrecht) [inline src/app_service/service/contacts/service.nim:148]: Also here

File: src/app_service/service/contacts/service.nim [hunk 6]
```diff
@@ -60,129 +71,99 @@ QtObject:
...
 
   proc blockContact*(self: Service, publicKey: string) =
-    var contact = self.getContact(publicKey)
+    var contact = self.getContactById(publicKey)
     contact.blocked = true
+
     self.saveContact(contact)
-    self.events.emit("contactBlocked", old_status_contacts.ContactIdArgs(id: publicKey))
+    self.events.emit(SIGNAL_CONTACT_BLOCKED, ContactArgs(contactId: contact.id))
```
Comment (PascalPrecht) [inline src/app_service/service/contacts/service.nim:155]: And here

File: src/app_service/service/contacts/service.nim [hunk 7]
```diff
@@ -60,129 +71,99 @@ QtObject:
...
+    var contact = self.getContactById(publicKey)
     contact.added = false
     contact.hasAddedUs = false
 
     self.saveContact(contact)
-    self.events.emit("contactRemoved", Args())
-  #   let channelId = status_utils.getTimelineChatId(publicKey)
-  #   if status_chat.hasChannel(channelId):
-  #     status_chat.leave(channelId)
+    self.events.emit(SIGNAL_CONTACT_REMOVED, ContactArgs(contactId: contact.id))
```
Comment (PascalPrecht) [inline src/app_service/service/contacts/service.nim:163]: and here

File: ui/app/AppMain.qml [hunk 8]
```diff
@@ -592,13 +593,22 @@ Item {
...
 
         Connections {
-            target: profileModel.contacts
+            target: profileModel
+            ignoreUnknownSignals: true
+            enabled: removeMnemonicAfterLogin
+            onInitialized: {
+                mnemonicModule.remove()
+            }
+        }
```
Comment (PascalPrecht) [inline ui/app/AppMain.qml:602]: Surprised to see this change in this commit. Is this still related to contacts refactor?

**Review by PascalPrecht — CHANGES_REQUESTED** [review]
Hey @alexandraB99 !

Massive work here. I think this is all a big improvement over what's currently in `master/base_02`.

I did notice that there's a problem though. Because `base_02` isn't rebased on top of latest `master` yet (and therefore this PR), the code still relies on contact APIs that can't be used anymore...

I'm referring to this commit here: https://github.com/status-im/status-desktop/pull/4152/files

Notice that we no longer call `saveContact` for every action, but instead call an API that performs the action for us. We have to make sure that this PR takes those changes into account as well.

**Comment by PascalPrecht** [comment]
@alexandraB99 I've updated `base_02` so you'll have to rebase this PR again (sorry! But we need to ensure we're using the correct APIs)

**Comment by jrainville** [comment]
Just an FYI @PascalPrecht , the reason this PR si so big is because it is the aggregation of 4 commits that come from base_bc that we done by Sale and I.
It's basically all the code to refactor contacts to the new architecture and remove the old code

**Comment by alexandraB99** [comment]
I have rebased this and it should have all 4 commits @jrainville mentioned [here](https://github.com/status-im/status-desktop/issues/4100#issuecomment-976649632). Nevertheless I still see an issue, which is in `ChatRequestMessagePanel` where it's not immediately reflected when a contact is added or removed but only when switching to char or any other view and going back to the contact view. Was this a know issue with your changes @jrainville?
https://user-images.githubusercontent.com/31625338/143307946-3f95194f-96df-4269-89eb-492a5193b89c.mov

**Comment by alexandraB99** [comment]
> Just an FYI @PascalPrecht , the reason this PR is so big is because it is the aggregation of 4 commits that come from base_bc that we done by Sale and I.
> It's basically all the code to refactor contacts to the new architecture and remove the old code

@jrainville could you please review and check whether all changes are in and correct and/if anything is missing or broken? ^^

**Comment by jrainville** [comment]
> I still see an issue, which is in ChatRequestMessagePanel where it's not immediately reflected when a contact is added or removed but only when switching to char or any other view and going back to the contact view. Was this a know issue with your changes @jrainville?

As far as I remember, that was working correctly for me. But then, there was one additional commit on top of mine, so maybe something was broken. 

I'll check the code to see if I see it

**Comment by jrainville** [comment]
Ok, so I checked so diffs, commits and the mainframe, and I think the problem is that Sale removed the call to `updateContactList` on new message event containing a contact list.

Check here: https://github.com/status-im/status-desktop/commit/4b3e5c6e36f41c382314d459cee6e762c7a5d43e#diff-bf318c2d799f85cf3cfea2b24889def7e1a817d9e699fbc0350130de11208382L37-L41

In case the link doesn't work, it's `src/app/modules/main/profile_section/contacts/controller.nim` lines 37-41.

If you try to re-add those lines (and fix it if ever the functions or events changed, does it work?

**Comment by alexandraB99** [comment]
> Ok, so I checked so diffs, commits and the mainframe, and I think the problem is that Sale removed the call to `updateContactList` on new message event containing a contact list.
> 
> Check here: [4b3e5c6#diff-bf318c2d799f85cf3cfea2b24889def7e1a817d9e699fbc0350130de11208382L37-L41](https://github.com/status-im/status-desktop/commit/4b3e5c6e36f41c382314d459cee6e762c7a5d43e#diff-bf318c2d799f85cf3cfea2b24889def7e1a817d9e699fbc0350130de11208382L37-L41)
> 
> In case the link doesn't work, it's `src/app/modules/main/profile_section/contacts/controller.nim` lines 37-41.
> 
> You can test first base_bc if the bug happens, then, if you try to re-add those lines (and fix it if ever the functions or events changed, does it work?

actually in base_bc the app is crashing when trying to add or remove a contact ^^
I tried adding these lines:

```
  self.events.on(SignalType.Message.event) do(e: Args):
    let msgData = MessageSignal(e);
    if msgData.contacts.len > 0:
      let contacts = self.getContacts(false)
      self.delegate.updateContactList(contacts)
```

in `init` method and I got `contacts/controller.nim(34, 18) Error: undeclared identifier: 'SignalType'`

**Comment by jrainville** [comment]
> I got contacts/controller.nim(34, 18) Error: undeclared identifier: 'SignalType'

You have to import it from `status/signals` if I'm not mistaken. It's from `status-lib`

**Comment by alexandraB99** [comment]
> status/signals

alright, so added

```
  self.events.on(SignalType.Message.event) do(e: Args):
    let msgData = MessageSignal(e);
    if msgData.contacts.len > 0:
      let contacts = self.getContacts()
      self.delegate.updateContactList(contacts)
```

and still no change

**Comment by jrainville** [comment]
That sucks. I can test some stuff later. I'll just finish the clean up of my other PR first

**Comment by jrainville** [comment]
@alexandraB99 I figured out the issue. It was related to the code snippet I put above, but it needed a bit more. The code was a bit faulty.
Anyway, it should work correctly now. You can check my commit to see.

**Comment by alexandraB99** [comment]
> @alexandraB99 I figured out the issue. It was related to the code snippet I put above, but it needed a bit more. The code was a bit faulty.
> Anyway, it should work correctly now. You can check my commit to see.

@jrainville I still see the issue on my end. I pulled your commit, made clean and build again and the contact view is not updating when in `ChatRequestMessagePanel`

**Comment by jrainville** [comment]
Ah it's possible that there was more than one issue.
I thought the issue was that receiving contact requests didn't update until restarting the app. So I fixed that one.
Can you describe and give repro steps for the issue you're having?

**Comment by alexandraB99** [comment]
> Ah it's possible that there was more than one issue.
> I thought the issue was that receiving contact requests didn't update until restarting the app. So I fixed that one.
> Can you describe and give repro steps for the issue you're having?

@jrainville sure thing, here you go:

https://user-images.githubusercontent.com/31625338/143497619-244e104f-0d83-42de-b576-c5c9f6994804.mov

**Comment by jrainville** [comment]
@alexandraB99 ok now I fully fixed it I think.

There were two problems, first, the 1-1 view didn't update on change. I added the change on the Connections
Second, Pascal changed the `save_contact` function to only do `add_contact`, so I changed the service to call the right status-lib functions.

This depends on this status-lib PR https://github.com/status-im/status-lib/pull/118

**Comment by alexandraB99** [comment]
thanks @jrainville! @PascalPrecht this is ready to review now.

**Review by jrainville — APPROVED** [review]

**Comment by PascalPrecht** [comment]
This works now @alexandraB99 !

Just pointing out that https://github.com/status-im/status-lib/pull/118 has to land first for this.
Also, the PR is conflicting right now with that dependency.

**Review by PascalPrecht — APPROVED** [review]

**Comment by alexandraB99** [comment]
> This works now @alexandraB99 !
> 
> Just pointing out that [status-im/status-lib#118](https://github.com/status-im/status-lib/pull/118) has to land first for this. Also, the PR is conflicting right now with that dependency.

ok, as soon as @jrainville merges that I would rebase.

**Comment by jrainville** [comment]
@alexandraB99 I merged. You can update the vendor
